src/ott/neural/methods/expectile_neural_dual.py

```diff
@@ -0,0 +1,526 @@
+        "cost_fn": self.cost_fn,
+        "is_bidirectional": self.is_bidirectional,
+        "corr": self._corr
+    }
+
+
+class PotentialModelWrapper(BasePotential):
+  """Wrapper class for the neural models.
+
+  Implements a potential value or a vector field.
+
+  Args:
+    model: Network architecture of the potential.
+    add_l2_norm: If `True``, l2 norm is added to the potential.
+    is_potential: Model the potential if ``True``, otherwise
+      model the gradient of the potential.
+  """
+
+  model: nn.Module
+  add_l2_norm: bool
+  is_potential: bool = True
+
+  @nn.compact
+  def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
+    """Apply model and optionally add l2 norm or x."""
+    z: jnp.ndarray = self.model(x)
+
+    if self.is_potential:
+      z = z.squeeze()
+
+    if self.add_l2_norm:
+      z = z + (0.5 * jnp.dot(x, x)) if self.is_potential else x
+
+    return z
+
+  def potential_gradient_fn(
+      self, params: FrozenDict[str, jnp.ndarray]
+  ) -> PotentialGradientFn_t:
+    """A vector function or gradient of the potential."""
+    if self.is_potential:
+      return jax.grad(self.potential_value_fn(params))
+    return lambda x: self.apply({"params": params}, x)
+
+
+class MLP(nn.Module):
+  """A simple MLP model of a potential used in default initialization."""
+
+  dim_hidden: Sequence[int]
+  act_fn: Callable[[jnp.ndarray], jnp.ndarray] = jax.nn.elu
+
+  @nn.compact
+  def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
+    """Apply MLP transform."""
+    for feat in self.dim_hidden[:-1]:
+      x = self.act_fn(nn.Dense(feat)(x))
+    return nn.Dense(self.dim_hidden[-1])(x)
+
+
+class ExpectileNeuralDual:
+  r"""Expectile-regularized Neural Optimal Transport (ENOT) :cite:`buzun:24`.
+
+  It solves the dual optimal transport problem for a specified cost function
+  c(x, y) between two measures :math:`\alpha` and :math:`\beta` in
+  :math:`d`-dimensional Euclidean space with additional regularization on
+  Kantorovich potentials. The expectile regularization enforces binding
+  conditions on the learning dual potentials :math:`f` and :math:`g`.
+  The main optimization objective is
+
+  .. math::
+
+    \sup_{g \in L_1(\beta)} \inf_{T: \, R^d \to R^d} \big[
+      \mathbb{E}_{\alpha}[c(x, T(x))] + \mathbb{E}_{\beta} [g(y)]
+      - \mathbb{E}_{\alpha} [g(T(x))]
+    \big],
+
+  where :math:`T(x)` is the transport mapping from :math:`\alpha`
+  to :math:`\beta` expressed through :math:`\nabla f(x)`. The explicit
+  formula depends on the cost function and ``is_bidirectional`` training
+  option. The regularization term is
+
+  .. math::
+
+    \mathbb{E} \mathcal{L}_{\tau} \big(
+      c(x, T(x)) - g(T(x)) - c(x, y)  + g(y)
+    \big),
+
+  where :math:`\mathcal{L}_{\tau}` is the least asymmetrically weighted
+  squares loss from expectile regression.
+
+  The :class:`flax.linen.Module` potentials for ``neural_f``
+  and ``neural_g`` can
+
+  1. both provide the values of the potentials :math:`f` and :math:`g`, or
```

Comment:
Please mention explicitly this is controlled by `is_bidirectional`.

Reply:
done